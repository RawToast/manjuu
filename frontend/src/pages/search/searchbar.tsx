import {
  BuildingOfficeIcon,
  BuildingOffice2Icon,
  HomeIcon,
  HomeModernIcon
} from '@heroicons/react/24/outline'
import {
  Command,
  CommandEmpty,
  CommandGroup,
  CommandInput,
  CommandItem,
  CommandList,
  CommandSeparator
} from '@/components/ui/command'
import { Authority, fetchAuthorities } from '@/lib/client'
import React from 'react'
import { useQuery } from '@tanstack/react-query'

export function SearchBar() {
  const { status, data, error } = useQuery({ queryKey: ['todos'], queryFn: fetchAuthorities })

  const [searchText, setSearchText] = React.useState('')

  if (status === 'error') {
    return <span>Error: {error.message}</span>
  }

  return (
    <Command className='rounded-lg border shadow-md'>
      <CommandInput
        placeholder='Type a command or search...'
        value={searchText}
        onValueChange={setSearchText}
      />
      {searchText.length == 0 ? (
        <></>
      ) : (
        <CommandList>
          <CommandEmpty>No results found.</CommandEmpty>

          <CommandGroup heading='Authorities'>
            {status === 'pending' ? (
              <CommandEmpty>Loading...</CommandEmpty>
            ) : (
              data.map((authority) => {
                return <AuthorityItem key={authority.id} authority={authority}></AuthorityItem>
              })
            )}
          </CommandGroup>

          <CommandSeparator />
          <CommandGroup heading='Recent'>
            <AuthorityItem
              authority={{ id: 10, name: 'Matlock', establishments: 65 }}
            ></AuthorityItem>
          </CommandGroup>
        </CommandList>
      )}
    </Command>
  )
}

function AuthorityItem({ authority }: { authority: Authority }) {
  return (
    <CommandItem key={authority.id}>
      {authority.establishments > 3000 ? (
        <BuildingOffice2Icon className='mr-2 h-4 w-4' />
      ) : authority.establishments > 2000 ? (
        <BuildingOfficeIcon className='mr-2 h-4 w-4' />
      ) : authority.establishments > 1000 ? (
        <HomeModernIcon className='mr-2 h-4 w-4' />
      ) : (
        <HomeIcon className='mr-2 h-4 w-4' />
      )}
      <span>{authority.name}</span>
      {/* <CommandShortcut>âŒ˜1</CommandShortcut> */}
    </CommandItem>
  )
}
