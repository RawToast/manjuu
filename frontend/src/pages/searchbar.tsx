import {
  BuildingOfficeIcon,
  BuildingOffice2Icon,
  HomeIcon,
  HomeModernIcon
} from '@heroicons/react/24/outline'
import {
  Command,
  CommandEmpty,
  CommandGroup,
  CommandInput,
  CommandItem,
  CommandList,
  CommandSeparator
} from '@/components/ui/command'
import { Authority, fetchAuthorities } from '@/lib/client'
import { useState } from 'react'
import { useQuery } from '@tanstack/react-query'
import { Link } from '@tanstack/react-router'

export function SearchBar() {
  const { status, data } = useQuery({ queryKey: ['todos'], queryFn: fetchAuthorities })

  const [searchText, setSearchText] = useState('')
  const [recent, setRecent] = useState<Authority[]>([])

  function updateRecent(authority: Authority) {
    // place on head of array, remove duplicates, and limit to 5
    setRecent(
      [authority, ...recent]
      // .filter((item, index, self) => self.findIndex((i) => i.id === item.id) === index)
      // .slice(0, 5)
    )
  }

  function onAuthorityClick(_e: React.MouseEvent<'a', MouseEvent>, _authorityId: number) {
    // Could update recent here
    setSearchText('')
  }

  return (
    <Command
      id='searchbar-label'
      className='rounded-lg border shadow-md'
      label='Search for an authority'
    >
      <CommandInput
        placeholder='Search for an authority...'
        value={searchText}
        onValueChange={(s) => {
          setSearchText(s)
        }}
        data-testid='searchbar-input'
      />
      {searchText.length <= 1 ? (
        <></>
      ) : (
        <CommandList>
          <CommandEmpty data-testid='search-no-results'>No results found.</CommandEmpty>
          <CommandGroup data-testid='search-authorities-group' heading='Authorities'>
            {/* Status error will bubble up to the top level */}
            {status === 'pending' && <CommandEmpty>Loading...</CommandEmpty>}
            {status === 'success' &&
              data.map((authority) => {
                return (
                  <Link
                    data-testid={`result-${authority.name.toLowerCase().replaceAll(' ', '-')}`}
                    key={'link' + authority.id}
                    to='/authority/$authorityId'
                    params={{ authorityId: `${authority.id}` }}
                    onClick={(e) => onAuthorityClick(e, authority.id)}
                    role='link'
                    title={authority.name}
                  >
                    <AuthorityItem
                      key={'item' + authority.id}
                      authority={authority}
                      onClick={updateRecent}
                    ></AuthorityItem>
                  </Link>
                )
              })}
          </CommandGroup>
          <CommandSeparator />
          {/* <CommandGroup heading='Recent'>
            {recent.map((authority) => (
              <AuthorityItem
                key={authority.id}
                authority={authority}
                onClick={updateRecent}
              ></AuthorityItem>
            ))}
          </CommandGroup> */}
        </CommandList>
      )}
    </Command>
  )
}

function AuthorityItem({
  authority,
  onClick
}: {
  authority: Authority
  onClick: (Authority) => void
}) {
  return (
    <CommandItem
      data-testid={`result-${authority.name.toLowerCase().replaceAll(' ', '-')}`}
      value={authority.name}
      key={authority.id}
      onClick={() => onClick(authority)}
    >
      {authority.establishments > 3000 ? (
        <BuildingOffice2Icon className='mr-2 h-4 w-4' />
      ) : authority.establishments > 2000 ? (
        <BuildingOfficeIcon className='mr-2 h-4 w-4' />
      ) : authority.establishments > 1000 ? (
        <HomeModernIcon className='mr-2 h-4 w-4' />
      ) : (
        <HomeIcon className='mr-2 h-4 w-4' />
      )}
      <span>{authority.name}</span>
      {/* <CommandShortcut>âŒ˜1</CommandShortcut> */}
    </CommandItem>
  )
}
