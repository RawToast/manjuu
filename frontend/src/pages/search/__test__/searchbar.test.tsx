import { render, screen } from '@testing-library/react'
import renderer from 'react-test-renderer'
import { expect, test } from 'vitest'

import { SearchBar } from '../searchbar'
import { QueryClient, QueryClientProvider } from '@tanstack/react-query'

test('Renders', () => {
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        retry: false
      },
      mutations: {
        retry: false
      }
    }
  })

  const underTest = (
    <QueryClientProvider client={queryClient}>
      <SearchBar />
    </QueryClientProvider>
  )

  render(underTest)
  const childText = screen.getByPlaceholderText(/Type a command or search/i)
  expect(childText).toBeInTheDocument()

  expect(renderer.create(underTest).toJSON()).toMatchSnapshot()
})
