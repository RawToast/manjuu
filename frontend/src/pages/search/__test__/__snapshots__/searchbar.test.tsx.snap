// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Renders 1`] = `
<div
  className="flex h-full w-full flex-col overflow-hidden bg-popover text-popover-foreground rounded-lg border shadow-md"
  cmdk-root=""
  onKeyDown={[Function]}
>
  <label
    cmdk-label=""
    htmlFor=":r2:"
    id=":r1:"
    style={
      {
        "borderWidth": "0",
        "clip": "rect(0, 0, 0, 0)",
        "height": "1px",
        "margin": "-1px",
        "overflow": "hidden",
        "padding": "0",
        "position": "absolute",
        "whiteSpace": "nowrap",
        "width": "1px",
      }
    }
  />
  <div
    className="flex items-center border-b px-3"
    cmdk-input-wrapper=""
  >
    <svg
      className="mr-2 h-4 w-4 shrink-0 opacity-50"
      fill="none"
      height="15"
      viewBox="0 0 15 15"
      width="15"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clipRule="evenodd"
        d="M10 6.5C10 8.433 8.433 10 6.5 10C4.567 10 3 8.433 3 6.5C3 4.567 4.567 3 6.5 3C8.433 3 10 4.567 10 6.5ZM9.30884 10.0159C8.53901 10.6318 7.56251 11 6.5 11C4.01472 11 2 8.98528 2 6.5C2 4.01472 4.01472 2 6.5 2C8.98528 2 11 4.01472 11 6.5C11 7.56251 10.6318 8.53901 10.0159 9.30884L12.8536 12.1464C13.0488 12.3417 13.0488 12.6583 12.8536 12.8536C12.6583 13.0488 12.3417 13.0488 12.1464 12.8536L9.30884 10.0159Z"
        fill="currentColor"
        fillRule="evenodd"
      />
    </svg>
    <input
      aria-autocomplete="list"
      aria-controls=":r0:"
      aria-expanded={true}
      aria-labelledby=":r1:"
      autoComplete="off"
      autoCorrect="off"
      className="flex h-10 w-full rounded-md bg-transparent py-3 text-sm outline-none placeholder:text-muted-foreground disabled:cursor-not-allowed disabled:opacity-50"
      cmdk-input=""
      id=":r2:"
      onChange={[Function]}
      placeholder="Type a command or search..."
      role="combobox"
      spellCheck={false}
      type="text"
      value=""
    />
  </div>
</div>
`;
